### PR TITLE
Modified alert dashbaord creation logic

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
tendrl-bug-id: Tendrl/monitoring-integration#354

Signed-off-by: Gowthamshanmugasundaram <gshanmug@redhat.com>

### CODE REVIEW

**Comment by GowthamShanmugam** [comment]
Each sync it will compare collected resource alerting panels from grafana if anything unrelated or not present in newly collected resource then it is removed.

**Comment by GowthamShanmugam** [comment]
@r0h4n @shtripat @nthomas-redhat please review

File: tendrl/monitoring_integration/grafana/alert_dashboard.py [hunk 1]
```diff
@@ -104,161 +102,154 @@ def create_resource_dashboard(
...
-                    if str(integration_id) in target["target"] and str(
-                            resource) in target["target"]:
-                        if resource_type == "bricks":
-                            if hostname in target["target"]:
-                                flag = True
+    # Keeping rows which are match with newly collcted cluster details
+    alert_dashboard = copy.deepcopy(resource_json)
+    resource_json["dashboard"]["rows"] = []
+    new_resource = []
+    highest_panel_id = 1
```
Comment (shtripat) [inline tendrl/monitoring_integration/grafana/alert_dashboard.py]: `most_recent_panel_id`

File: tendrl/monitoring_integration/sync/dashbaord_sync.py [hunk 2]
```diff
@@ -47,118 +63,118 @@ def refresh_dashboard(self):
...
-                            alert_dashboard["dashboard"]["rows"]
+    def create_alert_dashboard(self, cluster_details):
+        for dashboard_name in cluster_details:
+            # check alert dashboard exist
+            resource_json = alert_utils.get_alert_dashboard(
+                dashboard_name
+            )
+            dashboard_found = True
+            if "message" in resource_json and \
+                    resource_json["message"] == "Dashboard not found":
```
Comment (shtripat) [inline tendrl/monitoring_integration/sync/dashbaord_sync.py]: I feel rather use `"Dashboard not found" in resource_json["message"]`

File: tendrl/monitoring_integration/sync/dashbaord_sync.py [hunk 3]
```diff
@@ -47,118 +63,118 @@ def refresh_dashboard(self):
...
+            # Create operation takes more excution time than update
+            # Create dashboard for one resource using that do update for
+            # other resources
+            resource = resources.pop()
+            resource_json = alert_dashboard.create_resource_dashboard(
+                resource_type,
+                resource,
+            )
+        # After creating first panel update the other panels
+        if resources:
```
Comment (shtripat) [inline tendrl/monitoring_integration/sync/dashbaord_sync.py]: Why this check again, we can have all the code below if condition at line#122

File: tendrl/monitoring_integration/grafana/alert_dashboard.py [hunk 4]
```diff
@@ -104,161 +102,154 @@ def create_resource_dashboard(
...
+                                panel_title,
+                                resource_name
+                            )
+                            panel["id"] = count
+                            panel["legend"]["show"] = False
+                            panel["title"] = panel["title"] + \
+                                " - " + str(new_title)
+                            count = count + 1
+                            panel_count = panel_count + 1
+                            if panel_count < 7:
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/grafana/alert_dashboard.py]: i think grafana can support 16 panels per line but it does not look good so i put a check to allow only 7 panels per row,  more panels in single row is not good to see

File: tendrl/monitoring_integration/grafana/alert_dashboard.py [hunk 5]
```diff
@@ -104,161 +102,154 @@ def create_resource_dashboard(
...
                        {'message': str(ex)})
 
 
-def add_panel(integration_id, resource_type, resource_name, sds_name):
-    resp = None
-    alert_dashboard = alert_utils.get_alert_dashboard(
-        resource_type
-    )
-    if alert_dashboard:
+def add_panel(resources, resource_type, alert_dashboard, highest_panel_id):
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/grafana/alert_dashboard.py]: i will change it

File: tendrl/monitoring_integration/grafana/alert_dashboard.py [hunk 6]
```diff
@@ -104,161 +102,154 @@ def create_resource_dashboard(
...
-                        )
-                        resp = alert_utils.post_dashboard(dash_json)
-            except Exception as ex:
-                logger.log("error", NS.get("publisher_id", None),
-                           {'message': "Error while updating "
-                            "dashboard for %s" % resource_name})
-                raise ex
-    return resp
+            if sds_name == constants.GLUSTER:
+                alert_rows = fetch_row(alert_dashboard)
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/grafana/alert_dashboard.py]: i will change it

File: tendrl/monitoring_integration/grafana/alert_dashboard.py (hunk 1 above)
Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/grafana/alert_dashboard.py]: ok

File: tendrl/monitoring_integration/sync/dashbaord_sync.py (hunk 2 above)
Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/sync/dashbaord_sync.py]: i will change it

File: tendrl/monitoring_integration/sync/dashbaord_sync.py (hunk 3 above)
Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/sync/dashbaord_sync.py]: only once i am allowing create a dashboard, so if any resource after creates also then i doing update dashbaord.

so let's take example only one resource in a list, so i am pop it from list and calling create a dashbaord function using that resource, now the list is empty so no need to call update dashbaord so in this check i am checking again after pop any resource is empty or has some resource

File: tendrl/monitoring_integration/grafana/alert_dashboard.py [hunk 7]
```diff
@@ -104,161 +102,154 @@ def create_resource_dashboard(
...
+    new_resource = []
+    highest_panel_id = 1
+    for resource in resources:
+        new_resource_flag = True
+        integration_id = str(resource["integration_id"])
+        for row in alert_dashboard["dashboard"]["rows"]:
+            if "panels" in row:
+                for target in row["panels"][0]["targets"]:
+                    resource_name = str(resource["resource_name"])
+                    if resource_type == "bricks":
```
Comment (nthomas-redhat) [inline tendrl/monitoring_integration/grafana/alert_dashboard.py]: What is the logic of this function. Why only the 'brick' resource type is mentioned here?

Comment (GowthamShanmugam) [reply tendrl/monitoring_integration/grafana/alert_dashboard.py]: Here i am identifying whether the panel is already existing for a particular resource. i am identifying by comparing resource name with a target in grafana resource panel json.

for volume, target is like tendrl.clusters.{cid}.olumes.{vol_name}.pcnt_used
so here checking volume name is enough, if it matches then i panel is already exists no need to create new one.

for hosts
tendrl.clusters.{cid}.nodes.{hostname}.memory.percent-buffered 
here comparing with hostname then panel  already  exists

but for brick
tendrl.clusters.{cid}.nodes.{hostname}.bricks.{brick_path}.percent-percent_bytes

so here i need to check both so i put if condition there it find whether type is brick or not, if yes fetch both and compare

**Comment by GowthamShanmugam** [comment]
@shtripat changes done

**Pull request merged**